File: src/runner/messages.rs
use serde_derive::{Deserialize, Serialize};

use crate::gameplay::game::OnGameStateUpdate;

#[derive(Deserialize, Serialize, Debug)]
#[serde(rename_all = "camelCase")]
pub enum UserMessage {
    NoUpdates,
    QueueUpRequest(QueueUpRequest),
    JoinLobbyRequest(JoinLobbyRequest),
    MovePlayerRequest(MovePlayerRequest),
}

#[derive(Deserialize, Serialize, Debug)]
#[serde(rename_all = "camelCase")]
pub enum ServerMessage {
    QueueUpResponse(Result<QueueUpResponse, String>),
    JoinLobbyResponse(Result<JoinLobbyResponse, String>),
    ServerPushUpdate(Option<ServerPushUpdate>),
    MovePlayerResponse(Result<MovePlayerResponse, String>),
}

#[derive(Deserialize, Serialize, Debug)]
#[serde(rename_all = "camelCase")]
pub enum ServerPushUpdate {
    PotentialMatchUpdate(PotentialMatchUpdate),
    MatchStatusChange(MatchStatusChange),
    GameStateChange(OnGameStateUpdate),
}

#[derive(Deserialize, Serialize, Debug)]
#[serde(rename_all = "camelCase")]
pub struct QueueUpRequest {
    pub nickname: String,
}

#[derive(Deserialize, Serialize, Debug)]
#[serde(rename_all = "camelCase")]
pub struct JoinLobbyRequest {
    pub match_id: String,
}

#[derive(Deserialize, Serialize, Debug)]
#[serde(rename_all = "camelCase")]
pub struct MovePlayerRequest {
    pub match_id: String,
    pub y_delta: i32,
}

#[derive(Deserialize, Serialize, Debug)]
#[serde(rename_all = "camelCase")]
pub struct QueueUpResponse {
    pub id: String,
}

#[derive(Deserialize, Serialize, Debug)]
#[serde(rename_all = "camelCase")]
pub struct JoinLobbyResponse {
    pub match_id: String,
}

#[derive(Deserialize, Serialize, Debug)]
#[serde(rename_all = "camelCase")]
pub struct PotentialMatchUpdate {
    pub match_id: String,
    pub opoonents_ids: Vec<PotentialPlayer>,
}

#[derive(Deserialize, Serialize, Debug, Clone)]
#[serde(rename_all = "camelCase")]
pub struct PotentialPlayer {
    pub id: String,
    pub nickname: String,
}

#[derive(Deserialize, Serialize, Debug)]
#[serde(rename_all = "camelCase")]
pub enum MatchStatusChange {
    Stop(String),
    Start(String, String),
    PlayerReady(String),
}

#[derive(Deserialize, Serialize, Debug)]
#[serde(rename_all = "camelCase")]
pub struct MovePlayerResponse {
    pub match_id: String,
}

File: pong-web/src/lib/messages.ts
export type UserMessage = {
    noUpdates?: NoUpdates
    queueUpRequest?: QueueUpRequest
    joinLobbyRequest?: JoinLobbyRequest;
    movePlayerRequest?: MovePlayerRequest;
}

export type ServerMessage = {
    queueUpResponse?:  QueueUpResponse
    joinLobbyResponse?: JoinLobbyResponse
    serverPushUpdate?: ServerPushUpdate
}

export type QueueUpRequest = {
    nickname: string
}

export type QueueUpResponse = {
    Ok: {id: string},
}

export type NoUpdates = 'noUpdates'

export type ServerPushUpdate = {
    potentialMatchUpdate?: PotentialMatchUpdate,
    matchStatusChange?: MatchStatusChange
    gameStateChange?: GameStateChange
}

export type PotentialMatchUpdate = {
    matchId: string,
    opoonentsIds: PotentialPlayer[]
}

export type PotentialPlayer = {
    id: string,
    nickname: string
}

export type JoinLobbyRequest = {
    matchId: string
}

export type JoinLobbyResponse = {
    Ok: {matchId: string}
}

export type MatchStatusChange = {
    start?: string[]
    stop?: string
    playerReady?: string
}

export type GameStateChange = {
    id: string,
    state: GameState,
    timestampMs: number,
}

export type GameState = {
    player1Pos: Player,
    player2Pos: Player,
    ballPos: BallInfo,
    countdown: number,
    score: Score,
    recentHandledActions: string[],
}

export type Player = {
    id: String,
    position: Position,
    dimensions: Dimensions,
}

export type MovePlayerRequest = {
    matchId: string,
    yDelta: number,
    actionId: string,
}

export type Position = { x: number; y: number };
export type  Dimensions = {0: number, 1: number};
export type MovementVector = {
    horizontalVector: number,
    verticalVector: number,
};

export type BallInfo = {
    position: Position,
    radius: number,
    movement: MovementVector
}

export type Movement = 'up' | 'down' | 'none'

export type Score = {
    leftPlayer: PlayerWithScore,
    rightPlayer: PlayerWithScore,
    winner: {some?: string},
}


export type PlayerWithScore = {
    player: string,
    score: number,
}